floors/floors.go

```diff
@@ -0,0 +1,135 @@
+package floors
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/mxmCherry/openrtb/v16/openrtb2"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	DEFAULT_DELIMITER      string = "|"
+	CATCH_ALL              string = "*"
+	SKIP_RATE_MIN          int    = 0
+	SKIP_RATE_MAX          int    = 100
+	MODEL_WEIGHT_MAX_VALUE int    = 100
+	MODEL_WEIGHT_MIN_VALUE int    = 0
+	ENFORCE_RATE_MIN       int    = 0
+	ENFORCE_RATE_MAX       int    = 100
+)
+
+type FloorConfig struct {
+	FloorEnabled      bool
+	EnforceRate       int
+	EnforceDealFloors bool
+}
+
+func NewFloorConfig(priceFloor config.PriceFloors) *FloorConfig {
+
+	floorConfig := FloorConfig{
+		FloorEnabled:      priceFloor.Enabled,
+		EnforceRate:       priceFloor.EnforceFloorsRate,
+		EnforceDealFloors: priceFloor.EnforceDealFloors,
+	}
+
+	return &floorConfig
+}
+
+func (fc *FloorConfig) Enabled() bool {
+	return fc.FloorEnabled
+}
+
+func (fc *FloorConfig) GetEnforceRate() int {
+	return fc.EnforceRate
+}
+
+func (fc *FloorConfig) EnforceDealFloor() bool {
+	return fc.EnforceDealFloors
+}
+
+type Floor interface {
+	Enabled() bool
+	GetEnforceRate() int
+	EnforceDealFloor() bool
+}
+
+// IsRequestEnabledWithFloor will check if floors is enabled in request
+func IsRequestEnabledWithFloor(Floors *openrtb_ext.PriceFloorRules) bool {
+
+	if Floors != nil && Floors.Enabled != nil && !*Floors.Enabled {
+		return *Floors.Enabled
+	}
+
+	return true
+}
+
+// UpdateImpsWithFloors will validate floor rules, based on request and rules prepares various combinations
+// to match with floor rules and selects appripariate floor rule and update imp.bidfloor and imp.bidfloorcur
+func UpdateImpsWithFloors(floorExt *openrtb_ext.PriceFloorRules, request *openrtb2.BidRequest, conversions currency.Conversions) []error {
+	var (
+		floorErrList      []error
+		floorModelErrList []error
+		floorVal          float64
+	)
+	floorData := floorExt.Data
+	if floorData == nil {
+		return floorModelErrList
+	}
+
+	floorModelErrList = validateFloorSkipRates(floorExt)
+	if len(floorModelErrList) > 0 {
+		return floorModelErrList
+	}
+
+	floorData.ModelGroups, floorModelErrList = validateFloorModelGroups(floorData.ModelGroups)
+	if len(floorData.ModelGroups) == 0 {
```

Comment:
Shouldn't this be checking if floorModelErrList is not empty?

Reply:
there is possibility that out of multiple modelGroups, some modelGroups can be invalid or some of the parameters can be invalid.
There can be case where in some model groups got rejected due to invalid parameters (i.e. Errors are present in floorModelErrList) and one or more model groups are valid, in this case we need to process valid model groups.
If all the model groups are rejected due to validation in that case we would be returning with floorModelErrList

Reply:
Wouldn't we also want to report validation errors for the model groups which were rejected? We could send them down as warnings.

Reply:
Added condition to check if at least one model group is present